pkg/controller/v1alpha1/autoscaler/keda.go

```diff
@@ -0,0 +1,183 @@
+package autoscalers
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/utils/pointer"
+
+	kedav1alpha1 "github.com/kedacore/keda/api/v1alpha1"
+
+	"errors"
+	"strings"
+
+	"github.com/go-logr/logr"
+	kedatype "github.com/kedacore/keda/pkg/generated/clientset/versioned/typed/keda/v1alpha1"
+	"github.com/oam-dev/kubevela/api/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func (r *AutoscalerReconciler) scaleByKEDA(scaler v1alpha1.Autoscaler, namespace string, log logr.Logger) error {
+	config := r.config
+	kedaClient, err := kedatype.NewForConfig(config)
+	if err != nil {
+		log.Error(err, "failed to initiate a KEDA client", "config", config)
+		return err
+	}
+
+	minReplicas := scaler.Spec.MinReplicas
+	maxReplicas := scaler.Spec.MaxReplicas
+	triggers := scaler.Spec.Triggers
+	scalerName := scaler.Name
+	targetWorkload := scaler.Spec.TargetWorkload
+
+	// kedaScalerFlag marks whether KEDA Scaler should be applied
+	var kedaScalerFlag = false
+
+	var kedaTriggers []kedav1alpha1.ScaleTriggers
+	for _, t := range triggers {
+		if t.Type == CronType {
```

Comment:
Can we write or **find** a packge for this cron related thing?  I hope to see enough tests for this package, as parse the time\duration\zone like this made me nervous.

Reply:
second that, the following 100 lines of code can be very tricky to test. Please break them up and test them individually at least.